File: backend/src/HIS.Core/Entities/AdrReport.cs
namespace HIS.Core.Entities;

/// <summary>
/// Phiếu báo cáo phản ứng có hại của thuốc (ADR) — Thông tư 51/2017/TT-BYT.
/// Module #5 #55 #56 #57 #58 #59.
/// Không FK cứng vào Patient/Prescription để tránh cascade; chỉ lưu Guid tùy chọn.
/// </summary>
public class AdrReport : BaseEntity
{
    // ─── Thông tin bệnh nhân ────────────────────────────────────────────────

    /// <summary>Tên bệnh nhân (free-text, không bắt buộc link Patient).</summary>
    public string PatientName { get; set; } = string.Empty;

    /// <summary>Mã bệnh nhân (tuỳ chọn).</summary>
    public string? PatientCode { get; set; }

    /// <summary>Tuổi bệnh nhân (năm hoặc tháng — ghi tự do).</summary>
    public string PatientAge { get; set; } = string.Empty;

    /// <summary>Giới tính: 1=Nam, 2=Nữ, 0=Không rõ.</summary>
    public int Gender { get; set; }

    /// <summary>Cân nặng (kg), tuỳ chọn.</summary>
    public decimal? Weight { get; set; }

    // ─── Liên kết đơn thuốc (tuỳ chọn, không FK navigation) ────────────────

    /// <summary>Id đơn thuốc liên quan (tuỳ chọn, không FK cứng).</summary>
    public Guid? PrescriptionId { get; set; }

    // ─── Thông tin thuốc nghi ngờ ───────────────────────────────────────────

    /// <summary>Tên thuốc nghi ngờ gây phản ứng (bắt buộc).</summary>
    public string DrugName { get; set; } = string.Empty;

    /// <summary>Liều dùng (VD: "500mg x 3 lần/ngày").</summary>
    public string? DrugDose { get; set; }

    /// <summary>Đường dùng (VD: "uống", "tiêm tĩnh mạch").</summary>
    public string? DrugRoute { get; set; }

    // ─── Mô tả phản ứng ─────────────────────────────────────────────────────

    /// <summary>Mô tả chi tiết phản ứng có hại (bắt buộc).</summary>
    public string ReactionDescription { get; set; } = string.Empty;

    /// <summary>Ngày/giờ khởi phát phản ứng.</summary>
    public DateTime ReactionStartDate { get; set; }

    /// <summary>Mức độ nghiêm trọng: 1=Nhẹ, 2=Trung bình, 3=Nặng, 4=Nguy kịch/Tử vong.</summary>
    public int Severity { get; set; }

    /// <summary>Kết quả/hậu quả (hồi phục hoàn toàn, di chứng, tử vong,...).</summary>
    public string? Outcome { get; set; }

    // ─── Xử trí & đánh giá ──────────────────────────────────────────────────

    /// <summary>Xử trí đã thực hiện (ngừng thuốc, điều trị hỗ trợ,...).</summary>
    public string? ManagementTaken { get; set; }

    /// <summary>Đánh giá quan hệ nhân quả (chắc chắn, có khả năng, có thể, không chắc).</summary>
    public string? Causality { get; set; }

    // ─── Thông tin báo cáo ──────────────────────────────────────────────────

    /// <summary>Tên người báo cáo.</summary>
    public string? ReporterName { get; set; }

    /// <summary>Ngày báo cáo.</summary>
    public DateTime ReportDate { get; set; }

    /// <summary>Ghi chú bổ sung.</summary>
    public string? Notes { get; set; }
}

File: frontend/src/modules/quality/api/adrReport.ts
import apiClient from '../../../services/apiClient';

const BASE = '/adr-report';

// ─── DTO interfaces ──────────────────────────────────────────────────────────

export interface AdrReportDto {
  id: string;
  patientName: string;
  patientCode?: string;
  patientAge: string;
  gender: number;          // 0=Không rõ, 1=Nam, 2=Nữ
  weight?: number;
  prescriptionId?: string;
  drugName: string;
  drugDose?: string;
  drugRoute?: string;
  reactionDescription: string;
  reactionStartDate: string; // ISO datetime
  /** 1=Nhẹ, 2=Trung bình, 3=Nặng, 4=Nguy kịch/Tử vong */
  severity: number;
  outcome?: string;
  managementTaken?: string;
  causality?: string;
  reporterName?: string;
  reportDate: string; // ISO datetime
  notes?: string;
  createdAt?: string;
}

export interface AdrReportSummaryDto {
  totalReports: number;
  severityMild: number;
  severityModerate: number;
  severitySevere: number;
  severityCritical: number;
  fromDate?: string;
  toDate?: string;
}

// ─── API calls ───────────────────────────────────────────────────────────────

export const getAdrReports = (params?: {
  from?: string;
  to?: string;
  keyword?: string;
}) =>
  apiClient
    .get<AdrReportDto[]>(BASE, { params })
    .then(r => r.data);

export const saveAdrReport = (dto: Partial<AdrReportDto>) =>
  apiClient
    .post<AdrReportDto>(BASE, dto)
    .then(r => r.data);

export const deleteAdrReport = (id: string) =>
  apiClient
    .delete(`${BASE}/${id}`)
    .then(r => r.data);

export const getAdrReportSummary = (params?: { from?: string; to?: string }) =>
  apiClient
    .get<AdrReportSummaryDto>(`${BASE}/report`, { params })
    .then(r => r.data);
